providers/oidc.go

```diff
@@ -207,13 +210,18 @@ func (p *OIDCProvider) createSessionStateInternal(ctx context.Context, rawIDToke
 	newSession.Email = claims.UserID // TODO Rename SessionState.Email to .UserID in the near future
 
 	newSession.User = claims.Subject
+	newSession.Groups = claims.Groups
 	newSession.PreferredUsername = claims.PreferredUsername
 
 	verifyEmail := (p.UserIDClaim == emailClaim) && !p.AllowUnverifiedEmail
 	if verifyEmail && claims.Verified != nil && !*claims.Verified {
 		return nil, fmt.Errorf("email in id_token (%s) isn't verified", claims.UserID)
 	}
 
+	if len(p.RequiredGroups) > 0 && !claims.containsGroups(p.RequiredGroups) {
+		return nil, fmt.Errorf("user groups claim does not contain any of the required groups (%s)", p.RequiredGroups)
```

Comment:
Let me dig a bit for you to get you a recommendation. I think this is the wrong place to do the authorization check. This error will be bubbled up and handled here: https://github.com/oauth2-proxy/oauth2-proxy/blob/2318716a8992686a31f4fadaa1ec3b12ff2dd22b/oauthproxy.go#L720

I don't think a 500 Internal Server error is a appropriate, we want a 403 permission denied.

I'll check how the other providers that have a concept of group based authorization handled this for you.

Reply:
Good point, I had only put it here because the existing email verification check exists here too, which also appeared to throw a 500, but happy to make it "proper".

Reply:
You'll want to implement a `ValidateGroup` method on the OIDC provider to do this.

https://github.com/oauth2-proxy/oauth2-proxy/blob/2318716a8992686a31f4fadaa1ec3b12ff2dd22b/providers/providers.go#L17

The google provider has a good example of doing group validation.

Reply:
Ah cool had missed that hook! will look into it.

Reply:
@NickMeves any objection to changing the `ValidateGroup` function to taking the session as a paramter? should then make this possible.

Reply:
Oof -- didn't realize how all over the place group membership was in the other providers.  It looks like google was the only provider that implemented that `ValidateGroup(string)` method and it used the `userinfo` endpoint to get the details (hence it didn't need a session).

I'll need to think about this more -- wherever this logic happens, it needs to happen after session creation during the authZ stages of the application flow.

Otherwise If I was once authorized and get a session, and you change the allowed groups and I should no longer be authorized, my existing valid session will fly by unless you check my group membership for each request.

Reply:
I just got through changing validate group to take the session but good point, will hold off pushing!

Reply:
`ValidateSessionState` is the spot for this -- that will be called on every proxied request during the `getAuthenticatedSession` method in oauthproxy.go.

Reply:
FYI - we are working out what the future of more fine-grained authorization logic would look like in this PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/685

In that world, AuthZ checks on post session conditions (like groups) would likely happen here https://github.com/oauth2-proxy/oauth2-proxy/blob/2318716a8992686a31f4fadaa1ec3b12ff2dd22b/oauthproxy.go#L791

So since `group` is now a globally available session field -- maybe we leave the OIDC specific area and do the groups check here? 🤔

Reply:
Moved to `ValidateSessionState` sans testing, taking a break but will get back to this later tonight/tomorrow.

Reply:
You're too quick 🤣 

So upon thinking about this more -- now that groups are globally available in the SessionState (even though you only implemented in OIDC) -- the group authorization check needs to be global.

The AuthZ group membership check you did should be in this `Proxy` function after the authentication check but before the proxy to the upstream.

This then makes us want to rename the `oidc-required-groups` flag to be more global.  Since `required-groups` naming gives the impression the logic is all or nothing AND like we were chatting about, I think a better name might be something like: `--allowed-groups`

Reply:
This way actual getting the the backend will require a membership in an allowed group -- but other things that look at the session (like the `/oauth2/userinfo` endpoint) still work and give session details back to the user without failing because they don't have the allowed-groups.

Reply:
Cool gotcha, so to summarize:

1. Rename oidc-required-groups to allowed-groups
2. Rename oidc-groups-claim to groups-claim
3. Move check into the `Proxy` function after auth but before the proxy call

Reply:
(1) Yep
(2) No, this is still specific to the OIDC implementation.
(3) Yes (and one more)

Its poorly named (if we are about to AuthZ too), but this logic would need to happen here as well: https://github.com/oauth2-proxy/oauth2-proxy/blob/2318716a8992686a31f4fadaa1ec3b12ff2dd22b/oauthproxy.go#L769

Users that hook into oauth2-proxy via an nginx `auth_request` directive or Kubernetes ingress auth annotation hit this endpoint via an nginx subrequest rather than have `oauth2-proxy` as an inline proxy in the HTTP flow.

We'll want to support authZ in both supported deployment architectures.

Reply:
Ok :pray: I think I got this right, still missing tests will get to those.